File: recursion/powerset.js
// time: O(2^n * n)
// space: O(2^n * n)
// where n is the length of the array
function powerset(array) {
	const subsets = [[]]
	for (const elem of array) {
		const length = subsets.length
		for (let i = 0; i < length; i++) {
			subsets.push(subsets[i].concat(elem))
		}
	}
	return subsets
}

powerset([1, 2, 3]) // [[], [1], [2], [3], [1, 2], [1, 3], [2, 3], [1, 2, 3]]

File: algo-expert/medium/recursion/powerset.py
# time: O(n * 2^n)
# space: O(n * 2^n)
# where n is the length of the input array
def powersetIterative(array):
    subsets = [[]]
    for elem in array:
        for i in range(len(subsets)):
            current = subsets[i]
            subsets.append(current + [elem])
    return subsets

def powersetRecursive(array, i = None):
    if i is None:
        i = len(array) - 1
    if i < 0:
        return [[]]
    elem = array[i]
    subsets = powersetRecursive(array, i - 1)
    for i in range(len(subsets)):
        currentSubset = subsets[i]
        subsets.append(currentSubset + [elem])
    return subsets

print(powersetIterative([1, 2, 3])) # [[], [1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]]
